csrc/executor.cpp

```diff
@@ -645,14 +645,325 @@ std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShapeOfOutput(
+      dims.emplace_back(inner_dim);
+      while (i < tensor_.dim()) {
+        if (i != outer_dim && i != inner_dim) {
+          dims.emplace_back(i);
+        }
+        i++;
+      }
+      tensor_ = tensor_.permute(dims);
+    }
+    std::vector<int64_t> new_shape;
+    for (auto i : c10::irange(tensor_.dim())) {
+      if (i == left) {
+        new_shape.emplace_back(-1);
+      } else if (i != left + 1) {
+        new_shape.emplace_back(tensor_.size(i));
+      }
+    }
+    tensor_ = tensor_.view(new_shape);
+    // update frontier
+    if (inner_dim < outer_dim) {
+      *inner_it = in;
+      frontier_.erase(outer_it);
+    } else {
+      *outer_it = in;
+      frontier_.erase(inner_it);
+    }
+  }
+
+  // Backward traverse split from allocation to rFactor. Needs to, for example,
+  // view tensor with shape [..., 15, ...] as [..., 3, 5, ...]
+  void handle(Merge* merge) {
+    auto out = merge->out();
+    auto inner = merge->inner();
+    auto outer = merge->outer();
+    auto factor = ee_.evaluate(inner->extent())->as<int64_t>();
+    auto out_it = std::find(frontier_.begin(), frontier_.end(), out);
+    // TORCH_INTERNAL_ASSERT(out_it != frontier_.end());
+    if (out_it == frontier_.end()) {
+      // TODO: see [Allocation domain on both side of rFactor]
+      return;
+    }
+    // view tensor
+    int64_t dim = std::distance(frontier_.begin(), out_it);
+    std::vector<int64_t> new_shape;
+    for (auto i : c10::irange(tensor_.dim())) {
+      if (i == dim) {
+        new_shape.emplace_back(-1);
+        new_shape.emplace_back(factor);
+      } else {
+        new_shape.emplace_back(tensor_.size(i));
+      }
+    }
+    tensor_ = tensor_.view(new_shape);
+    // update frontier
+    frontier_.insert(out_it, outer);
+    frontier_.insert(out_it, inner);
+    frontier_.erase(out_it);
+  }
+
+  void handle(Expr* expr) {
+    if (auto split = dynamic_cast<Split*>(expr)) {
+      handle(split);
+    } else if (auto merge = dynamic_cast<Merge*>(expr)) {
+      handle(merge);
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          false, "Unsupported transormation in allocation domain");
+    }
+  }
+
+ public:
+  BackwardTraverseFromAllocToRFactor(
+      at::Tensor tensor,
+      TensorView* tv,
+      ExpressionEvaluator& ee,
+      std::list<IterDomain*>& frontier)
+      : tensor_(std::move(tensor)), tv_(tv), ee_(ee), frontier_(frontier) {}
+
+  at::Tensor run(
+      const std::vector<IterDomain*>& rfactor,
+      const std::vector<IterDomain*>& alloc) {
+    auto backward_exprs = StmtSort::getExprsBetween(
+        tv_->fusion(),
+        {rfactor.begin(), rfactor.end()},
+        {alloc.begin(), alloc.end()});
+    std::reverse(backward_exprs.begin(), backward_exprs.end());
+    for (auto expr : backward_exprs) {
+      handle(expr);
+    }
+    return tensor_;
+  }
+};
+
+// Start from a tensor whose dimensions are consistent with the allocation
+// domain of tv, transform view/permute the tensor into a format whose
```

Comment:
Can you elaborate what is exactly transformed here? It doesn't do any data copy, does it?

Reply:
No, there is no data copy, just apply a sequence of view and permute to the original tensor to view it as the desired format.

Reply:
Thanks. So, this is necessary because we may have an allocation domain that may be different from what the user would expect?

Reply:
Yes, the users is always expecting rFactor domain, which can be different from allocation domain.